[JIT] Add EliminateExceptions and RemoveInplaceOps to PE pass pipeline.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54732 [JIT] Add EliminateExceptions and RemoveInplaceOps to PE pass pipeline.**
* #54730 [JIT] Add EliminateExceptions pass.

Differential Revision: [D27343163](https://our.internmc.facebook.com/intern/diff/D27343163)

Commit:
[JIT] Add EliminateExceptions and RemoveInplaceOps to PE pass pipeline.

[ghstack-poisoned]

eellison reviewed (commented):
RemoveInplaceOps is the ONNX version of RemoveMutation that doesn't test correctness. I dont think we can turn it on because it might break model semantics. 

Even with testing correctness, when I tried out `RemoveMutation` (the one that does preserve semantics) I didnt see a speed up. As you mentioned, there are many cases like Conv-Relu-Conv-Relu where uninplacing the relu_ will lead to perf degradation. When I added [reinplacing](https://github.com/pytorch/pytorch/pull/53908) to MKLDNN subgraphs it sped things up by 16%, so I do think it kind of matters. 

What is the motivation for this ? If this is for an AOT workflow could we just have that workflow call RemoveTensorMutation ?

ZolotukhinM commented:
> ...there are many cases like Conv-Relu-Conv-Relu where uninplacing the relu_ will lead to perf degradation

Yeah, this change is questionable. I put it here because it does help to fuse more - for instance for mobilenet we start to include hardtanh and relu with this change. But I agree with your arguments and I'm ok with not running it here, in case it's needed for AOT I definitely can run it there manually.

Are you ok with the `EliminateExceptions` addition?

eellison commented:
> Are you ok with the EliminateExceptions addition?

Yea sure, just so long as it is not on by default! But if it's part of the AOT use case might make more sense there than in PE, not sure yet

Krovatkin commented on torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:
```diff
@@ -442,6 +461,11 @@ void ProfilingGraphExecutorImpl::runProfilingOptimizations(
     return;
   }
 
+  if (!exceptionsEnabled()) {
```
if this needs to be on the profiling path, maybe a small comment what this pass is for.

Commit:
Update on "[JIT] Add EliminateExceptions and RemoveInplaceOps to PE pass pipeline."


Differential Revision: [D27343163](https://our.internmc.facebook.com/intern/diff/D27343163)

[ghstack-poisoned]
